Add documentation on making dev overlay plugins (#5183)
withastro/docs: src/content/docs/en/reference/dev-overlay-reference.mdx

```ts title="src/my-plugin.js" {3-6}
  export default {
    // ...
    init(canvas, eventTarget) {
      eventTarget.addEventListener('plugin-toggle', (event) => {
        if (event.detail.state === true) {
          console.log("The plugin is now enabled!")
        }
      })
    }
  }
```

### `plugin-notification`

This event can be sent to inform the user that the plugin requires attention.

```ts title="src/my-plugin.js" {3-6}
  export default {
    // ...
    init(canvas, eventTarget) {
      eventTarget.dispatchEvent(
          new CustomEvent('plugin-notification', {
            detail: {
              state: true,
            },
          })
        );
    }
  }
```

`state` can be either `true` or `false`. When `true`, the plugin icon will be highlighted using a red dot. Conversely, when `false`, the highlight will be removed.

## Client-Server Communication

Using [Vite's methods for client-server communication](https://vitejs.dev/guide/api-plugin.html#client-server-communication), dev overlay plugins can communicate with the server.

> In addition to being able to send and receive arbritary messages, the dev overlay also sends the following messages, where `PLUGIN_ID` is the ID of a plugin:

Reviewer: ```suggestion
In addition to being able to send and receive custom messages, the dev overlay also sends the following messages, where `PLUGIN_ID` is the [plugin's ID](#id-string):
```
I'm not sure "custom" is the right word, but "arbitrary" sounds a lot like "random" so I'm wondering if we can choose something that more means "messages you've written for your plugin"...

> ... your plugin's messages ...

?

### `astro-dev-overlay:PLUGIN_ID:init`

This message is sent when the plugin is initialized. The data for this message is empty.

```ts title="integration.ts"

```

### `astro-dev-overlay:PLUGIN_ID:toggle`

> This message is sent when the user clicks on the plugin icon in the UI. The data for this message is a boolean indicating whether the plugin is enabled or not.

Reviewer: ```suggestion
This message is sent when the user clicks on the plugin icon in the UI. The data for this message is [the plugin's state](#state-boolean) indicating whether the plugin is enabled or not.
```
Would this be correct? This might tie it back nicely to the property!

Also, if we're showing an empty message above, is it worth showing the message being either true or false here, too?

> :::note

Reviewer: Just a note that I want to come back to this note!

The built-in `connection` event from Vite fire *before* dev overlay plugins are initialized. As such, trying to receive events from this event will not work inside plugins. Instead, use the `astro-dev-overlay:PLUGIN_ID:init` event.
:::

## UI Toolkit

The dev overlay includes a set of webcomponents that can be used to build plugins with a consistent look and feel.

### `astro-dev-overlay-window`

> Shows a window with optionally a title and an icon. `windowIcon` can either be a string of a SVG file or an icon from [the icon list](/en/reference/dev-overlay-plugin-api/#icons).

Reviewer: ```suggestion
Shows a window with optionally a title and an icon.

`windowTitle` is a string that will be shown at the top of the overlay window. `windowIcon` can either be a string of a SVG file or an icon from [the icon list](#icons).
```
I agree this maybe feels stupidly unnecessary, but it is a part of the API and it should be mentioned explicitly somewhere.

The slot of the component will be used as the content of the window.

```html
<astro-dev-overlay-window windowTitle="MyWindows" windowIcon="astro:logo">
  <p>My content</p>
</astro-dev-overlay-window>
```

### `astro-dev-overlay-card`

> Shows a card with a icon. Optionally, if `link` is passed, the card will be clickable and will open the link in a new tab.

Reviewer: ```suggestion
Shows a card with an [`icon`](#icons). Optionally, if a `link` is passed, the card will be clickable and will open the link in a new tab.
```
We have a section, so I like linking to it!

```

### `astro-dev-overlay-highlight`

Can be used to highlight an element on the page. In most cases, you'll want to position and resize this element using the `top`, `left`, `width` and `height` CSS properties to match the element you want to highlight. An icon can also be specified using the `icon` attribute and will be shown in the top right corner of the highlight.

```html
<!-- Highlight the entire page -->
<astro-dev-overlay-highlight style="top: 0; left: 0; width: 100%; height: 100%;"></astro-dev-overlay-highlight>
```

```ts
const elementToHighlight = document.querySelector('h1');
const rect = elementToHighlight.getBoundingClientRect();

const highlight = document.createElement('astro-overlay-highlight');

highlight.style.top = `${Math.max(rect.top + window.scrollY - 10, 0)}px`;
highlight.style.left = `${Math.max(rect.left + window.scrollX - 10, 0)}px`;
highlight.style.width = `${rect.width + 15}px`;
highlight.style.height = `${rect.height + 15}px`;
highlight.icon = 'astro:logo';
```

### `astro-dev-overlay-tooltip`

Shows a tooltip with different sections. This component is set to `display: none;` by default and can be made visible using a `data-show="true"` attribute.

> Sections are defined using the `sections` property. This property is an array of objects with the following shadpw:

Author: I have no idea what happened.
